src/common/hist_builder.h

```diff
@@ -0,0 +1,156 @@
+  const uint32_t* offsets = gmat.index.Offset();
+  const size_t n_features = row_ptr[1] - row_ptr[0];
+  const uint32_t two {2};  // Each element from 'gpair' and 'hist' contains
+                           // 2 FP values: gradient and hessian.
+                           // So we need to multiply each row-index/bin-index by 2
+                           // to work with gradient pairs as a singe row FP array
+  std::vector<std::vector<FPType>>& hists = *p_hists;
+
+  for (size_t i = row_begin; i < row_end; ++i) {
+    const size_t ri = is_root ? i : rows[i];
+    const size_t icol_start = any_missing ? row_ptr[ri] : ri * n_features;
+    const size_t icol_end =  any_missing ? row_ptr[ri+1] : icol_start + n_features;
+    const size_t row_size = icol_end - icol_start;
+    const size_t idx_gh = two * ri;
+    const uint32_t nid = is_root ? 0 : mapping_ids[nodes_ids[ri]];
+
+    if (do_prefetch) {
+      const size_t icol_start_prefetch = any_missing ? row_ptr[rows[i+Prefetch::kPrefetchOffset]] :
+                                       rows[i + Prefetch::kPrefetchOffset] * n_features;
+      const size_t icol_end_prefetch = any_missing ?  row_ptr[rows[i+Prefetch::kPrefetchOffset]+1] :
+                                      icol_start_prefetch + n_features;
+
+      PREFETCH_READ_T0(pgh + two * rows[i + Prefetch::kPrefetchOffset]);
+      for (size_t j = icol_start_prefetch; j < icol_end_prefetch;
+        j+=Prefetch::GetPrefetchStep<uint32_t>()) {
+        PREFETCH_READ_T0(gradient_index + j);
+      }
+    } else if (is_root) {
+      nodes_ids[ri] = 0;
+    }
+
+    const BinIdxType* gr_index_local = gradient_index + icol_start;
+    FPType* hist_data = hists[nid].data();
+
+    for (size_t j = 0; j < row_size; ++j) {
+      const uint32_t idx_bin = two * (static_cast<uint32_t>(gr_index_local[j]) + (
+                                      any_missing ? 0 : offsets[j]));
+      hist_data[idx_bin]   += pgh[idx_gh];
+      hist_data[idx_bin+1] += pgh[idx_gh+1];
+    }
+  }
+}
+
+/*!
+ * \brief builder for histograms of gradient statistics
+ */
+template<typename GradientSumT>
+class GHistBuilder {
+ public:
+  using GHistRowT = GHistRow<GradientSumT>;
+  GHistBuilder() = default;
+  GHistBuilder(size_t nthread, uint32_t nbins) : nthread_{nthread}, nbins_{nbins} {}
+
+  // construct a histogram via histogram aggregation
+  template <typename BinIdxType, bool any_missing, bool is_root>
+  void BuildHist(const std::vector<GradientPair>& gpair,
+                 const uint32_t* rows,
+                 const uint32_t row_begin,
+                 const uint32_t row_end,
+                 const GHistIndexMatrix& gmat,
+                 const BinIdxType* numa_data,
+                 uint16_t* nodes_ids,
+                 std::vector<std::vector<GradientSumT>>* p_hists,
+                 const uint16_t* mapping_ids) {
+    const size_t nrows = row_end - row_begin;
+    const size_t no_prefetch_size = Prefetch::NoPrefetchSize(nrows);
+
+    if (is_root) {
```

Comment:
May I ask what's the difference between the `is_root` can existing `contiguousBlock`?

Reply:
contiguous block is mostly possible on the root node and it's really rare case when nodes with depth > 0 have contiguous blocks so there is no need to handle such cases specifically.
Also `is_root` parameter handles the reset to zero `node_ids` buffer. As described above we need to track the node ids for each row (rows are not grouped for each node) so we have to reset nod_ids buffer to zero at the start of each tree building